src/cets_join.erl

```diff
@@ -38,29 +71,35 @@ join_loop(LockKey, Info, LocalPid, RemotePid, Start) ->
         %% overloaded or joining is already in progress on another node
         ?LOG_INFO(Info#{what => join_got_lock, after_time_ms => Diff}),
         %% Do joining in a separate process to reduce GC
-        cets_long:run_spawn(Info, fun() -> join2(Info, LocalPid, RemotePid) end)
+        cets_long:run_spawn(Info, fun() -> join2(Info, LocalPid, RemotePid, JoinOpts) end)
     end,
     LockRequest = {LockKey, self()},
     %% Just lock all nodes, no magic here :)
     Nodes = [node() | nodes()],
     Retries = 1,
     case global:trans(LockRequest, F, Nodes, Retries) of
         aborted ->
+            run_step(before_retry, JoinOpts),
             ?LOG_ERROR(Info#{what => join_retry, reason => lock_aborted}),
-            join_loop(LockKey, Info, LocalPid, RemotePid, Start);
+            join_loop(LockKey, Info, LocalPid, RemotePid, Start, JoinOpts);
         Result ->
             Result
     end.
 
-join2(_Info, LocalPid, RemotePid) ->
+join2(_Info, LocalPid, RemotePid, JoinOpts) ->
+    run_step(join_start, JoinOpts),
```

Comment:
For me it would be cleaner if there was a `lists:foldl` over a list of functions performing the actual steps, and in tests you could use `meck` to inject some logic when needed - by changing the actual step function.